File: 22_专题/离线查询/根号分治/PointSetStepSum.ts
/* eslint-disable no-inner-declarations */

// P3396 哈希冲突
// https://www.luogu.com.cn/problem/P3396
// https://www.luogu.com.cn/blog/danieljiang/ha-xi-chong-tu-ti-xie-gen-hao-ke-ji

class PointSetStepSum {
  private readonly _nums: Float64Array
  private readonly _stepThreshold: number

  /**
   * dp[step][start] 表示步长为step,起点为start的所有元素的和.
   * `dp[step][start] = dp[step][start+step] + nums[start]`.
   */
  private readonly _dp: Float64Array[]

  /**
   * @param stepThreshold 步长阈值,当步长小于等于该值时,使用dp数组预处理答案,否则直接遍历.
   * 预处理时间空间复杂度均为`O(n*stepThreshold)`.
   * 单次遍历时间复杂度为`O(n/stepThreshold)`.
   */
  constructor(arr: ArrayLike<number>, stepThreshold = 80) {
    this._nums = new Float64Array(arr)
    this._stepThreshold = stepThreshold
    this._dp = Array(stepThreshold)
    const n = arr.length
    for (let step = 1; step <= stepThreshold; step++) {
      const curSum = new Float64Array(n + 1)
      for (let start = n - 1; ~start; start--) {
        curSum[start] = curSum[Math.min(n, start + step)] + arr[start]
      }
      this._dp[step - 1] = curSum
    }
  }

  set(index: number, value: number): void {
    if (index < 0 || index >= this._nums.length) return
    const pre = this._nums[index]
    if (pre === value) return
    this._nums[index] = value
    const delta = value - pre
    for (let step = 1; step <= this._stepThreshold; step++) {
      this._dp[step - 1][index % step] += delta
    }
  }

  query(start: number, step: number): number {
    if (start < 0) start = 0
    if (step <= this._stepThreshold) {
      return this._dp[step - 1][start]
    }
    let sum = 0
    for (let i = start; i < this._nums.length; i += step) {
      sum += this._nums[i]
    }
    return sum
  }

  toString(): string {
    return `PointSetStepSum{${this._nums.join(',')}}`
  }
}

export { PointSetStepSum }

if (require.main === module) {
  // 1714. 数组中特殊等间距元素的和
  // https://leetcode.cn/problems/sum-of-special-evenly-spaced-elements-in-array/
  function solve(nums: number[], queries: number[][]): number[] {
    const MOD = 1e9 + 7
    const R = new PointSetStepSum(nums)
    const res: number[] = Array(queries.length)
    for (let i = 0; i < queries.length; i++) {
      const { 0: start, 1: step } = queries[i]
      res[i] = R.query(start, step) % MOD
    }
    return res
  }
}

File: 22_专题/离线查询/根号分治/PointSetStepSum.go
// PointSetStepSum
// 注意 !start<step.

package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	luogu3396()
}

// P3396 哈希冲突
// https://www.luogu.com.cn/problem/P3396
// https://www.luogu.com.cn/blog/danieljiang/ha-xi-chong-tu-ti-xie-gen-hao-ke-ji
func luogu3396() {
	in := bufio.NewReader(os.Stdin)
	out := bufio.NewWriter(os.Stdout)
	defer out.Flush()

	var n, q int
	fmt.Fscan(in, &n, &q)
	nums := make([]int, n)
	for i := 0; i < n; i++ {
		fmt.Fscan(in, &nums[i])
	}

	S := NewPointSetStepSum(nums, 50)
	for i := 0; i < q; i++ {
		var op string
		fmt.Fscan(in, &op)
		if op == "A" {
			var step, start int
			fmt.Fscan(in, &step, &start)
			start--
			if start < 0 {
				start += step
			}
			fmt.Fprintln(out, S.Query(start, step))
		} else {
			var pos, target int
			fmt.Fscan(in, &pos, &target)
			pos--
			S.Set(pos, target)
		}
	}
}

type PointSetStepSum struct {
	nums          []int
	stepThreshold int
	// dp[step][mod] 表示步长为step时，模为mod的所有数之和.
	dp [][]int
}

// stepThreshold: 步长阈值,当步长小于等于该值时,使用dp数组预处理答案,否则直接遍历.
// 预处理时间复杂度为`O(n*stepThreshold)`, 空间复杂度为O(`stepThreshold^2`)
// 单次遍历时间复杂度为`O(n/stepThreshold)`.
// 取50较为合适.
func NewPointSetStepSum(nums []int, stepThreshold int) *PointSetStepSum {
	nums = append(nums[:0:0], nums...)
	n := len(nums)
	dp := make([][]int, stepThreshold)
	for step := 1; step <= stepThreshold; step++ {
		curSum := make([]int, step)
		for i := 0; i < n; i++ {
			curSum[i%step] += nums[i]
		}
		dp[step-1] = curSum
	}
	return &PointSetStepSum{nums: nums, stepThreshold: stepThreshold, dp: dp}
}

func (pss *PointSetStepSum) Set(index, value int) {
	if index < 0 || index >= len(pss.nums) {
		return
	}
	pre := pss.nums[index]
	if pre == value {
		return
	}
	pss.nums[index] = value
	delta := value - pre
	for step := 1; step <= pss.stepThreshold; step++ {
		pss.dp[step-1][index%step] += delta
	}
}

// 查询 sum(nums[start::step]).
// !start<step.
func (pss *PointSetStepSum) Query(start, step int) int {
	if start < 0 {
		start = 0
	}
	if step <= pss.stepThreshold {
		return pss.dp[step-1][start]
	}
	sum := 0
	for i := start; i < len(pss.nums); i += step {
		sum += pss.nums[i]
	}
	return sum
}

func (pss *PointSetStepSum) Get(index int) int {
	if index < 0 || index >= len(pss.nums) {
		return 0
	}
	return pss.nums[index]
}

func (pss *PointSetStepSum) Add(index, delta int) {
	if index < 0 || index >= len(pss.nums) {
		return
	}
	if delta == 0 {
		return
	}
	pss.nums[index] += delta
	for step := 1; step <= pss.stepThreshold; step++ {
		pss.dp[step-1][index%step] += delta
	}
}

func (pss *PointSetStepSum) String() string {
	return fmt.Sprintf("PointSetModSum{nums: %v}", pss.nums)
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}
